exonum/src/encoding/fields.rs

```diff
@@ -309,38 +319,66 @@ impl<'a> Field<'a> for DateTime<Utc> {
 impl<'a> Field<'a> for SocketAddr {
     fn field_size() -> Offset {
         // FIXME: reserve space for future compatibility (ECR-156)
-        6
+        (SOCKET_ADDR_HEADER_SIZE + IPV6_SIZE + PORT_SIZE) as Offset
     }
 
     unsafe fn read(buffer: &'a [u8], from: Offset, to: Offset) -> Self {
-        let mut octets = [0u8; 4];
-        octets.copy_from_slice(&buffer[from as usize..from as usize + 4]);
-        let ip = Ipv4Addr::from(octets);
-        let port = LittleEndian::read_u16(&buffer[from as usize + 4..to as usize]);
-        SocketAddr::V4(SocketAddrV4::new(ip, port))
+        let addr_start = from as usize + SOCKET_ADDR_HEADER_SIZE;
+        let ip = match buffer[from as usize] {
+            IPV4_HEADER => {
+                let mut octets: [u8; IPV4_SIZE] = mem::uninitialized();
+                octets.copy_from_slice(&buffer[addr_start..addr_start + IPV4_SIZE]);
+                IpAddr::V4(Ipv4Addr::from(octets))
+            }
+            IPV6_HEADER => {
+                let mut octets: [u8; IPV6_SIZE] = mem::uninitialized();
+                octets.copy_from_slice(&buffer[addr_start..addr_start + IPV6_SIZE]);
+                IpAddr::V6(Ipv6Addr::from(octets))
+            }
+            header => panic!("Unknown header `{:X}` for SocketAddr", header),
+        };
+        let port = LittleEndian::read_u16(&buffer[to as usize - PORT_SIZE..to as usize]);
+        SocketAddr::new(ip, port)
     }
 
     fn write(&self, buffer: &mut Vec<u8>, from: Offset, to: Offset) {
         match *self {
-            SocketAddr::V4(addr) => {
-                buffer[from as usize..to as usize - 2].copy_from_slice(&addr.ip().octets());
+            SocketAddr::V4(ref addr) => {
+                buffer[from as usize] = IPV4_HEADER;
```

Comment:
Does Rust or the `write` method provides any guarantee as to the uninitialized values of bytes in `buffer` (as in this case, with 12 uninitialized bytes after the IPv4 address)? Could it be a source of malleability / failed round trip? Maybe, it would make sense that these padding bytes are equal to 0 i the `check` method?